### PR TITLE
Feature/angular support

### PR DESCRIPTION
**Pull request by joaquinmarti**
Support for Angular CSF stories. It needs to build an entire app with its module and also a wrapper component to host the story component bindings.

### CODE REVIEW

File: src/render.ts [hunk 1]
```diff
@@ -133,6 +133,37 @@ export async function render(
       div.appendChild(storyResult);
       return true;
     }
+    case "Angular": {
+      const { props, component: StoryComponent, markup } = storyResult;
+      const { __decorate } = (await require("tslib"));
```
Comment (gqio) [inline src/render.ts]: do we really need this ?

File: src/render.ts [hunk 2]
```diff
@@ -133,6 +133,37 @@ export async function render(
       div.appendChild(storyResult);
       return true;
     }
+    case "Angular": {
```
Comment (gqio) [inline src/render.ts:136]: I think this one is a bit trickier. You should also provide also services maybe via story.context/args, don't you ?
Could you please check with the render I made within the calypso ds.

File: src/typeOf.ts [hunk 3]
```diff
@@ -38,6 +38,14 @@ export function typeOf(obj: any): string {
     if ("__v_isVNode" in obj || "__scopeId" in obj || ("components" in obj && ("template" in obj || "render" in obj))) {
       return "Vue";
     }
+    if (
```
Comment (gqio) [inline src/typeOf.ts:41]: nice 👍

File: src/render.ts (hunk 2 above)
Comment (joaquinmarti) [reply src/render.ts:136]: I believe you mean this, right?

```
    const {
      imports = [],
      declarations = [],
      bootstrap = [],
      providers = [],
      schemas = [],
    } = context.args;
    @NgModule({
      imports: [...dependencies, ...imports],
      declarations: [...declarations, storyComponent],
      bootstrap: [...bootstrap, storyComponent],
      providers,
      schemas,
    })
    ```

Comment (joaquinmarti) [reply src/render.ts:136]: Storybook does this:

```
import { moduleMetadata } from '@storybook/angular';
import { CommonModule } from '@angular/common';

export default {
  component: List,
  title: 'List',
  decorators: [
    moduleMetadata({
      declarations: [List,ListItem],
      imports: [CommonModule],
    }),
  ],
} as Meta;
```

File: src/render.ts (hunk 1 above)
Comment (joaquinmarti) [reply src/render.ts]: You are right, probably not as the render file is already TS.

Comment (joaquinmarti) [reply src/render.ts]: This should be solved after the last commit.

**Review by gqio — APPROVED** [review]
looks promising this one 👍

**Comment by joaquinmarti** [comment]
Still need to pass imports, bootstrap and declarations. Will try to do that similar to Storybook.

**Comment by joaquinmarti** [comment]
Ok @gqio. Please check the last changes. I am passing there the module metadata from the story. This has to be done through a decorator, following CSF specs. 

For now I am doing this:

```
export default {
  component: SimpleCounter,
  decorators: [
    (Story) => ({
      ...Story(),
      imports: [],
      declarations: [],
      bootstrap: [],
      providers: [],
      schemas: [],
    }),
  ],
};
```

Ideally we could do decorator to make it more syntactic sugar, similar to what Storybook does:

```
export default {
  component: ButtonGroup,
  decorators: [
    moduleMetadata({
      declarations: [Button],
      imports: [CommonModule],
    }),
  ],

```

But for that we'd need a studio package so we can import that decorator in the story (it could also host more helpers the studio need to build csf). I didn't do it for now to a void the overhead of publishing an npm package. Let me know what you think.

File: src/render.ts [hunk 4]
```diff
@@ -133,6 +133,42 @@ export async function render(
...
+      const { props, component: StoryComponent, markup } = storyResult;
+      const { platformBrowserDynamic } = (await require("@angular/platform-browser-dynamic"));
+      const { Component, NgModule, destroyPlatform } = (await require("@angular/core"));
+      const { BrowserModule } = (await require("@angular/platform-browser"));
+      const {
+        imports = [],
```
Comment (gqio) [inline src/render.ts]: I think it's not following what CSF needs/does. moduleMetadata shoulld be a field on the storyresult If I am reading correctly what they are doing right here https://unpkg.com/browse/@storybook/angular@6.3.12/dist/ts3.9/client/preview/decorators.js

Comment (gqio) [reply src/render.ts]: I think you can destructure the storyresult for the metadata.

Comment (joaquinmarti) [reply src/render.ts]: Correct, that's pretty easy to solve but then we'd probably need our own decorator so we don't force the user to write the decorator with the Story function and the moduleMetadata property.

```
export default {
  component: SimpleCounter,
  decorators: [
    (Story) => ({
      ...Story(),
      moduleMetadata: {
        imports: [],
        declarations: [],
        bootstrap: [],
        providers: [],
        schemas: [],
      }
    }),
  ],
};
```

Should we have a new package for this (and others!) Angular feature?

Comment (gqio) [reply src/render.ts]: no need to do that. since you already proposed something related around angular to csf

Comment (joaquinmarti) [reply src/render.ts]: Updated.

Comment (gqio) [reply src/render.ts]: lgtm

File: src/render.ts [hunk 5]
```diff
@@ -133,6 +133,42 @@ export async function render(
...
+        imports: [BrowserModule, ...imports],
+        declarations: [AppComponent, StoryComponent, ...declarations],
+        bootstrap: [AppComponent, ...bootstrap],
+        schemas,
+        providers,
+      })
+      class AppModule {};
+
+      // Init the app
+      div.insertAdjacentHTML('beforeend', '<app-root></app-root>');
```
Comment (gqio) [inline src/render.ts]: Sorry maybe I misunderstand this, but why adjacent ? Shouldn't we reuse the div already inplace ? the div where we want to render it ?

Comment (gqio) [reply src/render.ts]: Can't we have a selector for this particular div ?

Comment (joaquinmarti) [reply src/render.ts]: The `insertAdjacentHTML` method with the `beforeend` renders the string inside the div where it's invoked.

That div is the root, the one that holds the stories.

Comment (gqio) [reply src/render.ts]: Yes I know but I am wondering why we need another div behind the scene. I don't know how much is hard to have the correct selector for the App itself.

File: src/render.ts [hunk 6]
```diff
@@ -133,6 +133,42 @@ export async function render(
       div.appendChild(storyResult);
       return true;
     }
+    case "Angular": {
+      const { props, component: StoryComponent, markup } = storyResult;
```
Comment (gqio) [inline src/render.ts]: you can rename with `markup : template` as in it's used as `template` below.

File: src/render.ts (hunk 5 above)
Comment (gqio) [reply src/render.ts]: Could you try with removing the `selector` in the AppComponent, this `div.insetAdjacentHTML` and use `platformBrowserDynamic().bootstrapModule(AppModule, div);` where div is the mount point for the story ?

Comment (gqio) [reply src/render.ts]: I can take over this if you need, just ask ;)

**Review by gqio — APPROVED** [review]

**Review by gqio — APPROVED** [review]

**Pull request merged**